libtenzir/builtins/formats/xsv.cpp

```diff
@@ -306,102 +416,117 @@ auto parse_impl(generator<std::optional<std::string_view>> lines,
-      for (auto&& slice :
-           b.finish_as_table_slice(fmt::format("tenzir.{}", args.name))) {
-        co_yield std::move(slice);
-      }
+    for (auto& v : msb.yield_ready_as_table_slice()) {
+      co_yield std::move(v);
     }
+    auto line = *it;
     if (not line) {
-      if (last_finish != now) {
-        co_yield {};
-      }
+      co_yield {};
       continue;
     }
+    ++line_counter;
     if (line->empty()) {
       continue;
     }
     if (args.allow_comments && line->front() == '#') {
       continue;
     }
-    const auto single_value_delimiter
-      = (parsers::eoi | args.list_sep | args.field_sep);
-    const auto single_value_parser
-      = (parsers::lit{args.null_value} >> &single_value_delimiter)
-          .then([](std::string) {
-            return data{};
-          })
-        | (parsers::data >> &single_value_delimiter).with([](const data& d) {
-            return caf::visit(
-              []<class T>(const T&) {
-                return not detail::is_any_v<T, pattern, std::string, list,
-                                            record>;
-              },
-              d);
-          })
-        | (qqstring_value_parser >> &single_value_delimiter
-           | *(parsers::any - single_value_delimiter))
-            .then([](std::string str) {
-              return data{std::move(str)};
-            });
-    const auto value_parser = (single_value_parser % args.list_sep)
-                                .then([](std::vector<data> values) -> data {
-                                  TENZIR_ASSERT(not values.empty());
-                                  if (values.size() == 1) {
-                                    return std::move(values[0]);
-                                  }
-                                  return values;
-                                });
-    auto values_parser = (value_parser % args.field_sep);
-    auto values = std::vector<data>{};
-    if (not values_parser(*line, values)) {
-      diagnostic::warning("skips unparseable line")
-        .note("from `{}` parser", args.name)
-        .emit(ctrl.diagnostics());
-      continue;
-    }
-    auto generated_field_id = 0;
-    if (args.auto_expand) {
-      while (fields.size() < values.size()) {
-        auto name = fmt::format("unnamed{}", ++generated_field_id);
-        if (std::find(fields.begin(), fields.end(), name) == fields.end()) {
-          fields.push_back(name);
+    auto r = msb.record();
+    size_t field_idx = 0;
+    for (field_idx = 0; true; ++field_idx) {
+      if (line->empty()) {
+        if (field_idx < original_field_count) {
+          diagnostic::warning("{} parser found too few values in a line",
+                              args.name)
+            .note("line {} has {} values, but should have {} values",
+                  line_counter, field_idx, original_field_count)
+            .emit(ctrl.diagnostics());
+        }
+        break;
+      } else if (field_idx >= fields.size()) {
```

Comment:
Help me remember: Was this a conscious decision to warn if a value is missing, and if so, what was the reasoning?